x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java

```diff
@@ -217,6 +241,40 @@ synchronized void shutdown() {
         markAsCompleted();
     }
 
+    void persistStateToClusterState(DataFrameTransformState state,
+                                    ActionListener<PersistentTasksCustomMetaData.PersistentTask<?>> listener) {
+        updatePersistentTaskState(state, ActionListener.wrap(
+            success -> {
+                logger.debug("Successfully updated state for data frame transform [{}] to [{}]", transform.getId(), state.toString());
+                listener.onResponse(success);
+            },
+            failure -> {
+                auditor.warning(transform.getId(), "Failed to persist to state to cluster state: " + failure.getMessage());
+                logger.error("Failed to update state for data frame transform [" + transform.getId() + "]", failure);
+                listener.onFailure(failure);
+            }
+        ));
+    }
+
+    private boolean isIrrecoverableFailure(Exception e) {
+        return e instanceof IndexNotFoundException || e instanceof DataFrameConfigurationException;
+    }
+
+    synchronized void handleFailure(Exception e) {
+        if (isIrrecoverableFailure(e) || failureCount.incrementAndGet() > MAX_CONTINUOUS_FAILURES) {
+            String failureMessage = isIrrecoverableFailure(e) ?
+                "task encountered irrecoverable failure: " + e.getMessage() :
+                "task encountered more than " + MAX_CONTINUOUS_FAILURES + " failures; latest failure: " + e.getMessage();
```

Comment:
We may want to make `MAX_CONTINUOUS_FAILURES` a setting one day.

Reply:
I think more important than this (I would wait until there is a real demand to expose this), would be exponential wait times. Right now the scheduler fires quickly and therefore it takes only some seconds to breach the limit. This isn't very practical in a real world scenario where e.g. a temporary network outage happens.

But this problem is not part of this PR. This limit is good, we just need to fix the scheduling logic which is part of the scheduler re-work.